flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientSavepointTriggerTest.java

```diff
@@ -0,0 +1,340 @@
+			new TestSavepointTriggerHandler(triggerHandlerLogic),
+			new TestSavepointHandler(savepointHandlerLogic));
+	}
+
+	private static final class TestSavepointTriggerHandler extends TestHandler<SavepointTriggerRequestBody, TriggerResponse, SavepointTriggerMessageParameters> {
+
+		private final FunctionWithException<SavepointTriggerRequestBody, TriggerId, RestHandlerException> triggerHandlerLogic;
+
+		TestSavepointTriggerHandler(final FunctionWithException<SavepointTriggerRequestBody, TriggerId, RestHandlerException> triggerHandlerLogic) {
+			super(SavepointTriggerHeaders.getInstance());
+			this.triggerHandlerLogic = triggerHandlerLogic;
+		}
+
+		@Override
+		protected CompletableFuture<TriggerResponse> handleRequest(
+			@Nonnull HandlerRequest<SavepointTriggerRequestBody, SavepointTriggerMessageParameters> request,
+			@Nonnull DispatcherGateway gateway) throws RestHandlerException {
+
+			return CompletableFuture.completedFuture(new TriggerResponse(triggerHandlerLogic.apply(request.getRequestBody())));
+		}
+	}
+
+	private static class TestSavepointHandler extends TestHandler<EmptyRequestBody, AsynchronousOperationResult<SavepointInfo>, SavepointStatusMessageParameters> {
+
+		private final FunctionWithException<TriggerId, SavepointInfo, RestHandlerException> savepointHandlerLogic;
+
+		TestSavepointHandler(final FunctionWithException<TriggerId, SavepointInfo, RestHandlerException> savepointHandlerLogic) {
+			super(SavepointStatusHeaders.getInstance());
+			this.savepointHandlerLogic = savepointHandlerLogic;
+		}
+
+		@Override
+		protected CompletableFuture<AsynchronousOperationResult<SavepointInfo>> handleRequest(
+				@Nonnull HandlerRequest<EmptyRequestBody, SavepointStatusMessageParameters> request,
+				@Nonnull DispatcherGateway gateway) throws RestHandlerException {
+
+			final TriggerId triggerId = request.getPathParameter(TriggerIdPathParameter.class);
+			return CompletableFuture.completedFuture(AsynchronousOperationResult.completed(savepointHandlerLogic.apply(triggerId)));
+		}
+	}
+
+	private abstract static class TestHandler<R extends RequestBody, P extends ResponseBody, M extends MessageParameters> extends AbstractRestHandler<DispatcherGateway, R, P, M> {
+
+		private TestHandler(MessageHeaders<R, P, M> headers) {
+			super(
+				mockGatewayRetriever,
+				RpcUtils.INF_TIMEOUT,
+				Collections.emptyMap(),
+				headers);
+		}
+	}
+
+	private static TestRestServerEndpoint createRestServerEndpoint(
+			final RestServerEndpointConfiguration restServerEndpointConfiguration,
+			final AbstractRestHandler<?, ?, ?, ?>... abstractRestHandlers) throws Exception {
+		final TestRestServerEndpoint testRestServerEndpoint = new TestRestServerEndpoint(restServerEndpointConfiguration, abstractRestHandlers);
+		testRestServerEndpoint.start();
+		return testRestServerEndpoint;
+	}
+
+	private static class TestRestServerEndpoint extends RestServerEndpoint implements AutoCloseable {
```

Comment:
I would extract this class to avoid code duplication with the other `RestClusterClientTest`. Same with `TestHandler`.